### PR TITLE
Implementation dmd (with randomization)

### PR DESCRIPTION
**Pull request by hvhue**
~First implemetation of Dynamic Mode Decomposition and some simple tests using the parabolic and heat demos~

Deterministic DMD implementations have been added to pyMOR in #1424. This PR still contains an unmerged randomized version.

### CODE REVIEW

**Comment by TiKeil** [comment]
Hi @hvhue, you should rebase your branch on main such that the CI can start.

**Comment by renefritze** [comment]
The gitlab documentation build failures are due to missing references. That's not straight-forward to see in the log. 
https://zivgitlab.uni-muenster.de/pymor/pymor/-/jobs/547837#L5496

**Comment by TiKeil** [comment]
Now the failures in the docs build comes can be found here
https://zivgitlab.uni-muenster.de/pymor/pymor/-/jobs/549064#L4166, 
i.e. you should seprate authors with "and" and not with a comma :), see
https://tex.stackexchange.com/questions/343163/error-in-bib-entry-too-many-commas

File: src/pymor/algorithms/dmd.py [hunk 1]
```diff
@@ -0,0 +1,270 @@
...
+            Wk = Wk*EVALS_inv
+
+    return Wk, omega
+
+
+def rand_QB(A, target_rank=None, distribution='normal', oversampling=0, powerIterations=0):
+    """
+    randomisierte QB-Zerlegung
+
+    See Algorithm 3.1 in [EMKB19]_.
```
Comment (sdrave) [inline src/pymor/algorithms/dmd.py:126]:
```suggestion
    See Algorithm 3.1 in :cite:`EMKB19`.
```

File: src/pymor/algorithms/dmd.py [hunk 2]
```diff
@@ -0,0 +1,270 @@
...
+    return Q, B
+
+
+@defaults('svd_method', 'distribution')
+def rand_dmd(A, target_rank=None, dt=1, modes='exact', svd_method='qr_svd', distribution='normal',
+             oversampling=0, powerIterations=0, order=True):
+    """
+    Ranzomized Dynamic Mode Decomposition
+
+    See Algorithm 4.1 in [EMKB19]_.
```
Comment (sdrave) [inline src/pymor/algorithms/dmd.py:187]:
```suggestion
    See Algorithm 4.1 in :cite:`EMKB19`.
```

File: src/pymor/algorithms/dmd.py [hunk 3]
```diff
@@ -0,0 +1,270 @@
...
+        if modes == 'exact_scaled':
+            EVALS_inv = np.reciprocal(EVALS)
+            Wk = Wk*EVALS_inv
+
+    return Wk, omega
+
+
+def rand_QB(A, target_rank=None, distribution='normal', oversampling=0, powerIterations=0):
+    """
+    randomisierte QB-Zerlegung
```
Comment (sdrave) [inline src/pymor/algorithms/dmd.py:124]:
```suggestion
    Randomized QB decomposition.
```

File: src/pymor/algorithms/dmd.py [hunk 4]
```diff
@@ -0,0 +1,270 @@
...
+
+    See Algorithm 3.1 in [EMKB19]_.
+
+    Parameters
+    ----------
+    A  :
+        The |VectorArray| for which the randomized QB Decomposition is to be computed.
```
Comment (sdrave) [inline src/pymor/algorithms/dmd.py:131]:
```suggestion
        The |VectorArray| for which the randomized QB decomposition is to be computed.
```

File: src/pymor/algorithms/dmd.py [hunk 5]
```diff
@@ -0,0 +1,270 @@
...
+
+    target_rank  :  int
+        The desired rank for the decomposition. If None rank = len(A).
```
Comment (sdrave) [inline src/pymor/algorithms/dmd.py:134]:
```suggestion
        The desired rank for the decomposition. If `None`, rank = len(A).
```

File: src/pymor/algorithms/dmd.py [hunk 6]
```diff
@@ -0,0 +1,270 @@
...
+
+    distribution : str
+        Distribution used for the random projectionmatrix Omega. (`'normal'` or `'uniform'`)
```
Comment (sdrave) [inline src/pymor/algorithms/dmd.py:137]:
```suggestion
        Distribution used for the random sample matrix (`'normal'` or `'uniform'`).
```

(The name 'omega' was not defined in the docstring so we should not use it ..

File: src/pymor/algorithms/dmd.py [hunk 7]
```diff
@@ -0,0 +1,270 @@
...
+
+    oversampling : int
+        Oversamplingparameter. Number of extra columns of the projectionmatrix.
```
Comment (sdrave) [inline src/pymor/algorithms/dmd.py:140]:
```suggestion
        Oversampling parameter. Number of extra columns of the sample matrix.
```

Is the term 'projection matrix' used in the literature? If any of the matrices, I would call 'Q' the projection matrix, not omega.

File: src/pymor/algorithms/dmd.py [hunk 8]
```diff
@@ -0,0 +1,270 @@
...
+    elif modes == 'exact' or 'exact_scaled':
+        Wk = XR.lincomb(V.T).lincomb(Sigma_inv.T).lincomb(EVECS.T)
+        if modes == 'exact_scaled':
+            EVALS_inv = np.reciprocal(EVALS)
+            Wk = Wk*EVALS_inv
+
+    return Wk, omega
+
+
+def rand_QB(A, target_rank=None, distribution='normal', oversampling=0, powerIterations=0):
```
Comment (sdrave) [inline src/pymor/algorithms/dmd.py:122]: Camel case violates PEP8. (Surprisingly, the linter did not seem to catch that.) Please use `power_iterations` instead.

File: src/pymor/algorithms/dmd.py [hunk 9]
```diff
@@ -0,0 +1,270 @@
...
+        The desired rank for the decomposition. If None rank = len(A).
+
+    distribution : str
+        Distribution used for the random projectionmatrix Omega. (`'normal'` or `'uniform'`)
+
+    oversampling : int
+        Oversamplingparameter. Number of extra columns of the projectionmatrix.
+
+    powerIterations : int
+        Number of power Iterations.
```
Comment (sdrave) [inline src/pymor/algorithms/dmd.py:143]:
```suggestion
        Number of power iterations.
```

File: src/pymor/algorithms/dmd.py [hunk 10]
```diff
@@ -0,0 +1,270 @@
...
+
+
+    Returns
+    -------
+    Q :
+        |VectorArray| containig an approximate optimal Basis for the Image of the Inputmatrix A.
```
Comment (sdrave) [inline src/pymor/algorithms/dmd.py:149]:
```suggestion
        |VectorArray| containig an approximate basis for the image of the input matrix A.
```

Not sure what optimal would mean in that context. Any basis for the image would work equally well.

File: src/pymor/algorithms/dmd.py [hunk 11]
```diff
@@ -0,0 +1,270 @@
...
+        len(Q) = target_rank
+    B :
+        Numpy Array. Projection of the Input Matrix into the lower dimensional subspace.
```
Comment (sdrave) [inline src/pymor/algorithms/dmd.py:152]:
```suggestion
        Numpy array. Projection of the input matrix into the lower dimensional subspace.
```

File: src/pymor/algorithms/dmd.py [hunk 12]
```diff
@@ -0,0 +1,270 @@
...
+    """
+    Ranzomized Dynamic Mode Decomposition
+
+    See Algorithm 4.1 in [EMKB19]_.
+
+
+    Parameters
+    ----------
+    A  :
+        The |VectorArray| for which the DMD Modes are to be computed.
```
Comment (sdrave) [inline src/pymor/algorithms/dmd.py:193]:
```suggestion
        The |VectorArray| for which the DMD modes are to be computed.
```

File: src/pymor/algorithms/dmd.py [hunk 13]
```diff
@@ -0,0 +1,270 @@
...
+
+    target_rank : int
+        Number of DMD Modes to be computed. If None target_rank = len(A).
```
Comment (sdrave) [inline src/pymor/algorithms/dmd.py:196]:
```suggestion
        Number of DMD modes to be computed. If `None` target_rank = len(A).
```

File: src/pymor/algorithms/dmd.py [hunk 14]
```diff
@@ -0,0 +1,270 @@
...
+        - 'exact' : computes the exact dynamic modes, `Wk = XR * V * Sigma_inv * EVECS`.
+        - 'exact_scaled' : computes the exact dynamic modes,
+                           `Wk = (1/EVALS) * XR * V * Sigma_inv * EVECS`.
+
+    svd_method : str
+        Which SVD method from :mod:`~pymor.algorithms.svd_va` to use
+        (`'method_of_snapshots'` or `'qr_svd'`).
+
+    distribution : str
+        Distribution used for the randomized QB-Decomposition. (`'normal'` or `'uniform'`)
```
Comment (sdrave) [inline src/pymor/algorithms/dmd.py:213]:
```suggestion
        Distribution used for the randomized QB-Decomposition (`'normal'` or `'uniform'`).
```

File: src/pymor/algorithms/dmd.py [hunk 15]
```diff
@@ -0,0 +1,270 @@
...
+
+    oversampling : int
+        Oversamplingparameter. Number of extra columns of the projectionmatrix.
```
Comment (sdrave) [inline src/pymor/algorithms/dmd.py:216]:
```suggestion
        Oversampling parameter. Number of extra columns of the sample matrix.
```
See above.

File: src/pymor/algorithms/dmd.py [hunk 16]
```diff
@@ -0,0 +1,270 @@
...
+
+    Q = gram_schmidt(Y)[:target_rank]
+    B = Q.inner(A)
+
+    return Q, B
+
+
+@defaults('svd_method', 'distribution')
+def rand_dmd(A, target_rank=None, dt=1, modes='exact', svd_method='qr_svd', distribution='normal',
+             oversampling=0, powerIterations=0, order=True):
```
Comment (sdrave) [inline src/pymor/algorithms/dmd.py:183]: See above regarding `powerIterations`.

File: src/pymor/algorithms/dmd.py [hunk 17]
```diff
@@ -0,0 +1,270 @@
...
+        True: return modes sorted.
+
+
+    Returns
+    -------
+    Wk :
+        |VectorArray| containing the dynamic modes.
+
+    omega :
+        Numpy Array containig time scaled eigenvalues: `ln(l)/dt`.
```
Comment (sdrave) [inline src/pymor/algorithms/dmd.py:231]:
```suggestion
        Numpy array containig time scaled eigenvalues: `ln(l)/dt`.
```

File: src/pymordemos/parabolicDMD_TEST.py [hunk 18]
```diff
@@ -0,0 +1,360 @@
...
+    nt=50
+)
+grid = data['grid']
+print(grid)
+print()
+
+print('Solve ...')
+U = m.solve()
+
+# m.visualize(U, title='Solution of Parabolic Problem')
```
Comment (sdrave) [inline src/pymordemos/parabolicDMD_TEST.py:46]:
```suggestion
```

File: src/pymordemos/parabolicDMD_TEST.py [hunk 19]
```diff
@@ -0,0 +1,360 @@
...
+# ----- Testing DMD -----
+
+"""
+Calculate Extrapolation Error
+"""
+
+U_k = U[-1]
+U = U[:-1]
+
+alle_fehler_dmd = []
```
Comment (sdrave) [inline src/pymordemos/parabolicDMD_TEST.py:57]: Please use only English identifiers ..

File: src/pymordemos/parabolicDMD_TEST.py [hunk 20]
```diff
@@ -0,0 +1,360 @@
...
+fehler_r2dmd_EW = []
+fehler_r3dmd_EW = []
+fehler_r4dmd_EW = []
+fehler_r5dmd_EW = []
+fehler_r6dmd_EW = []
+
+# Testmoden = [1, 2, 3, 4, 5, 6, 7, 8, 9, 10, 11, 12, 13, 14, 15, 16, 17, 18, 19, 20]
+Testmoden = [1, 5, 10, 15, 20]
+# Testmoden = [1, 2, 3, 4, 5, 10]
+# Testmoden = [1, 20, 50, 80]
```
Comment (sdrave) [inline src/pymordemos/parabolicDMD_TEST.py:91]:
```suggestion
```

File: src/pymordemos/parabolicDMD_TEST.py [hunk 21]
```diff
@@ -0,0 +1,360 @@
...
+alle_zeiten_r6dmd = []
+
+fehler_rdmd_EW = []
+fehler_r2dmd_EW = []
+fehler_r3dmd_EW = []
+fehler_r4dmd_EW = []
+fehler_r5dmd_EW = []
+fehler_r6dmd_EW = []
+
+# Testmoden = [1, 2, 3, 4, 5, 6, 7, 8, 9, 10, 11, 12, 13, 14, 15, 16, 17, 18, 19, 20]
```
Comment (sdrave) [inline src/pymordemos/parabolicDMD_TEST.py:88]:
```suggestion
```

**Review by sdrave — CHANGES_REQUESTED** [review]
Hi @hvhue, thanks for your pull request! I added several minor comments. 

Regarding the demo scripts: Can you try to merge the test scripts and use typer for cli argument parsing as it is done in `parabolic.py`? I didn't check what `parabolicDMD_test` does compared to the other two scripts, but generally I would say that one script is enough. We should discuss at some point what a good testcase for dmd would be. We can still change the concrete example later on, however. 

Apart from the demo scripts it would also be great if we also had a short [tutorial](https://docs.pymor.org/2020.2.0/tutorials.html) which also very briefly explains the main mathematical ideas behind the method.

**Review by renefritze — CHANGES_REQUESTED** [review]
It would be great if you can add the demo script into the test suite, too. That is fairly straightforward to do by adding a new tuple of arguments [here](https://github.com/pymor/pymor/blob/main/src/pymortests/demos.py#L122) (first string is the demo name, rest demo args) and then adding that tuple to `DEMO_ARGS`.

File: src/pymor/algorithms/dmd.py [hunk 22]
```diff
@@ -0,0 +1,270 @@
...
+    """Dynamic Mode Decomposition.
+
+    See Algorithm 1 and Algorithm 2 in :cite:`TRLBK14`.
+
+    Parameters
+    ----------
+    A  :  optional
+        The |VectorArray| for which the DMD Modes are to be computed.
+        If A is used, XL and XR should not be used.
+
```
Comment (sdrave) [inline src/pymor/algorithms/dmd.py:25]: In pyMOR, we don't have blank lines between arguments. Could you remove them?

File: src/pymor/algorithms/dmd.py [hunk 23]
```diff
@@ -0,0 +1,270 @@
...
+        - 'exact_scaled' : computes the exact dynamic modes,
+                           `Wk = (1/EVALS) * XR * V * Sigma_inv * EVECS`.
+
+    order : bool `{True, False}`
+        True: return modes sorted.
+
+    svd_method
+        Which SVD method from :mod:`~pymor.algorithms.svd_va` to use
+        (`'method_of_snapshots'` or `'qr_svd'`).
+
```
Comment (sdrave) [inline src/pymor/algorithms/dmd.py:51]:
```suggestion
```

File: src/pymor/algorithms/dmd.py [hunk 24]
```diff
@@ -0,0 +1,270 @@
...
+    Returns
+    -------
+    Wk : VectorArray
+        |VectorArray| containing the dynamic modes.
+
+    omega : array_like
+        Time scaled eigenvalues: `ln(l)/dt`.
+    """
+
```
Comment (sdrave) [inline src/pymor/algorithms/dmd.py:61]:
```suggestion
```

File: src/pymor/algorithms/dmd.py [hunk 25]
```diff
@@ -0,0 +1,270 @@
...
+    rank = len(XL) if target_rank is None else target_rank
+    svd = qr_svd if svd_method == 'qr_svd' else method_of_snapshots
+
+    print('SVD of XL...')
+    U, SVALS, Vh = svd(XL, product=None, modes=rank)
+
+    # Invert the Singular Values
+    SVALS_diag = np.diag(SVALS)
+    Sigma_inv = np.linalg.inv(SVALS_diag)
```
Comment (sdrave) [inline src/pymor/algorithms/dmd.py:87]:
```suggestion
    Sigma_inv = np.diag(1/SVALS)
```

File: src/pymor/algorithms/dmd.py [hunk 26]
```diff
@@ -0,0 +1,270 @@
...
+    V = Vh.conj().T
+
+    # Solve the least Sqaures Problem
```
Comment (sdrave) [inline src/pymor/algorithms/dmd.py:90]:
```suggestion
    # Solve the least Squares Problem
```

File: src/pymor/algorithms/dmd.py [hunk 27]
```diff
@@ -0,0 +1,270 @@
...
+    # real: A_tilde = U.T * XR * Vh.T * Sigma_inv
+    # complex: A_tilde = U.H * XR * Vh.H * Sigma_inv
+    # A = XR[:k] @ V @ Sigma_inv.inner(Uh)
+    A_tilde = U.inner(XR) @ V @ Sigma_inv
+
+    print('Calculating eigenvalue dec. ...')
+    EVALS, EVECS = spla.eig(A_tilde, b=None, left=False, right=True)
+    # omega = np.log(EVALS) / dt
```
Comment (sdrave) [inline src/pymor/algorithms/dmd.py:98]: Why is this here?

File: src/pymor/algorithms/dmd.py [hunk 28]
```diff
@@ -0,0 +1,270 @@
...
+
+    # XL = x0, ..., xm-1  ; XR = x1, ..., xm
+    if A is not None:
+        XL = A[:-1]
+        XR = A[1:]
+
+    rank = len(XL) if target_rank is None else target_rank
+    svd = qr_svd if svd_method == 'qr_svd' else method_of_snapshots
+
+    print('SVD of XL...')
```
Comment (sdrave) [inline src/pymor/algorithms/dmd.py:82]: Can you replace all print statements in the library code by corresponding logging code like, for instance, in `pymor.algorithms.gram_schmidt.gram_schmidt`?

File: src/pymor/algorithms/dmd.py [hunk 29]
```diff
@@ -0,0 +1,270 @@
...
+        if modes == 'exact_scaled':
+            EVALS_inv = np.reciprocal(EVALS)
+            Wk = Wk*EVALS_inv
+
+    return Wk, omega
+
+
+def rand_QB(A, target_rank=None, distribution='normal', oversampling=0, power_iterations=0):
+    """
+    Ramdomized QB decomposition
```
Comment (sdrave) [inline src/pymor/algorithms/dmd.py:124]:
```suggestion
    """Ramdomized QB decomposition.
```

File: src/pymor/algorithms/dmd.py [hunk 30]
```diff
@@ -0,0 +1,270 @@
...
+    # transform Wk to numpy
+    Wk = Wk.to_numpy()
+
+    Wk = Q.lincomb(Wk)
+
+    return Wk, omega
+
+
+def get_amplitudes(A, Wk):
+    '''Compute amplitueds b using least-squares: Fb=x1'''
```
Comment (sdrave) [inline src/pymor/algorithms/dmd.py:263]:
```suggestion
    """Compute amplitueds b using least-squares: Fb=x1"""
```

File: src/pymor/algorithms/dmd.py [hunk 31]
```diff
@@ -0,0 +1,270 @@
...
+    W_OP = VectorArrayOperator(Wk)
+    return lsqr(W_OP, A[0])[0].to_numpy()[0]
+
+
+def get_vandermonde(A, EVals):
+    '''Compute Vandermonde matrix'''
```
Comment (sdrave) [inline src/pymor/algorithms/dmd.py:269]:
```suggestion
    """Compute Vandermonde matrix"""
```

File: src/pymordemos/burgers_dmd.py [hunk 32]
```diff
@@ -0,0 +1,141 @@
+"""Burgers demo with different Applications of Dynamic Mode Decomposition.
+
+"""
```
Comment (sdrave) [inline src/pymordemos/burgers_dmd.py:3]:
```suggestion
"""Burgers demo with different Applications of Dynamic Mode Decomposition."""
```

File: src/pymor/algorithms/dmd.py [hunk 33]
```diff
@@ -0,0 +1,270 @@
...
+
+    if A.dim == 0 or len(A) == 0:
+        return A.space.zeros(), np.zeros((target_rank, len(A)))
+
+    rank = len(A) if target_rank is None else target_rank + oversampling
+    target_rank = len(A) if target_rank is None else target_rank
+
+    Omega = np.random.normal(0, 1, (rank, len(A))) if distribution == 'normal' else np.random.rand(rank, len(A))
+
+    Y = A.lincomb(Omega)[:target_rank]
```
Comment (sdrave) [inline src/pymor/algorithms/dmd.py:166]: This doesn't look quite right! With the slicing you just throw away all additional vectors you get from oversampling!? I think there is actually an issue with the original paper. Maybe you should look at this point more closely and also take a look at what is done in HMT11. We can also discuss this further in more detail if you wish.

**Review by sdrave — CHANGES_REQUESTED** [review]
Hi @hvhue, this is looking much better already! Some more comments:

- Both demo scripts seem to do exactly the same, except for applying dmd to different FOMs. Maybe you can merge the two scripts into one?
- In `algorithms.randrangefinder`, there are already randomized algorithms to determine the range of an `Operator`. Maybe you could merge your code with the existing algorithms?
- You should add an entry to `AUTHORS.md`!

**Comment by sdrave** [comment]
I forgot: You also need to rebase on main to get CI working again.

**Comment by renefritze** [comment]
This was superseded by #1424 and can be closed, right @pymor/main ?

**Comment by sdrave** [comment]
I'm closing this now, as the PR is quite old now and of it is in #1424. The randomized part could be easily reimplemented and should use `pymor.algorithms.rand_la`.

**Pull request closed without merging**